allennlp/data/iterators/iterator.py

```diff
@@ -0,0 +1,24 @@
+from typing import Dict, List, Generator
+
+import numpy
+
+from ..dataset import Dataset
+from ...common.params import Params
+
+
+class Iterator:
+    """
+    An abstract Iterator class. Iterators must implement __call__, which yields
+    batched examples.
+    """
+    def __call__(self, dataset: Dataset) -> Generator[Dict[str, List[numpy.array]], None, None]:
```

Comment:
Those are definitely good reasons to prefer `__call__` over `__iter__`, it's more flexible.